numpyro/util.py

```diff
@@ -240,22 +142,33 @@ def fn_loop(n, x): return lax.fori_loop(0, n, body_fun, x)
     return tree_unflatten(trans_treedef, state)
 
 
-def fori_collect(n, body_fun, init_val, transform=_identity):
+def fori_collect(n, body_fun, init_val, transform=_identity, use_prims=True):
     # works like lax.fori_loop but ignores i in body_fn, supports
     # postprocessing `transform`, and collects values during the loop
     init_val_flat, unravel_fn = ravel_pytree(transform(init_val))
+    ravel_fn = lambda x: ravel_pytree(transform(x))[0]  # noqa: E731
 
-    @jit
-    def _body_fun(states):
-        val = body_fun(states[0])
-        return val, ravel_pytree(transform(val))[0]
+    if use_prims:
+        collection = np.zeros((n,) + init_val_flat.shape, dtype=init_val_flat.dtype)
+
+        def _body_fn(i, vals):
+            val, collection = vals
+            val = body_fun(val)
+            collection = ops.index_update(collection, i, ravel_fn(val))
+            return val, collection
+
+        _, collection = jit(lax.fori_loop, static_argnums=(2,))(0, n, _body_fn,
```

Comment:
How does this version compare with `fori_append`?

Reply:
The speed is the same as fori_append in test models.

Reply:
In that case, let us just profile on `test_mcmc.py` and remove `fori_append`, since there shouldn't be a reason to have both of these?